src/compiler/srcloc.rs

```diff
@@ -29,19 +45,102 @@ pub struct Srcloc {
 
 impl Display for Srcloc {
     fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
-        match self.until {
+        match &self.until {
             None => formatter.write_str(&format!("{}({}):{}", self.file, self.line, self.col)),
-            Some((l, c)) => formatter.write_str(&format!(
+            Some(u) => formatter.write_str(&format!(
                 "{}({}):{}-{}({}):{}",
-                self.file, self.line, self.col, self.file, l, c
+                self.file, self.line, self.col, self.file, u.line, u.col
             )),
         }
     }
 }
 
 impl Srcloc {
+    pub fn new(name: Rc<String>, line: usize, col: usize) -> Self {
+        Srcloc {
+            file: name,
+            line,
+            col,
+            until: None,
+        }
+    }
+
+    pub fn ending(&self) -> Srcloc {
+        if let Some(u) = &self.until {
+            return Srcloc {
+                file: self.file.clone(),
+                line: u.line,
+                col: u.col,
+                until: None,
+            };
+        }
+        self.clone()
+    }
+
+    pub fn overlap(&self, other: &Srcloc) -> bool {
+        let mf: &String = self.file.borrow();
+        let of: &String = other.file.borrow();
+        if mf != of {
+            return false;
+        }
+
+        if self.line == other.line && self.col == other.col {
+            return true;
+        }
+
+        match (self.until.as_ref(), other.until.as_ref()) {
+            (None, None) => self.line == other.line && self.col == other.col,
+            (None, Some(_)) => other.overlap(self),
+            (Some(u), None) => {
+                if self.line < other.line && u.line > other.line {
+                    return true;
+                }
+                if self.line == other.line
+                    && self.col <= other.col
+                    && self.col + self.len() >= other.col
+                {
+                    return true;
+                }
+                if u.line == other.line && self.col + self.len() >= other.col {
+                    return true;
+                }
+
+                false
+            }
+            (Some(u1), Some(u2)) => {
+                let l1 = Srcloc::new(self.file.clone(), self.line, self.col);
+                let l2 = Srcloc::new(self.file.clone(), u1.line, u1.col);
+                let l3 = Srcloc::new(self.file.clone(), other.line, other.col);
+                let l4 = Srcloc::new(self.file.clone(), u2.line, u2.col);
+                other.overlap(&l1) || other.overlap(&l2) || self.overlap(&l3) || self.overlap(&l4)
+            }
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        false
+    }
+
+    pub fn len(&self) -> usize {
```

Comment:
Is `len` == number of chars(bytes?) from start to until?

Should probably put a comment in here.

Reply:
Made it into Option<usize> so it's more clear.